lib/history/sqlite.go

```diff
@@ -0,0 +1,379 @@
+		}
+	}
+	return nil
+}
+
+// evictEvents deletes oldest events if the timeline is larger than its max
+// capacity.
+func (t *SQLiteTimeline) evictEvents(ctx context.Context, tx *sql.Tx) error {
+	if _, err := tx.ExecContext(ctx, deleteOldFromEvents, t.capacity); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// prepareQuery prepares a query string and a list of arguments constructed from
+// the provided params.
+func prepareQuery(params map[string]string) (query string, args []interface{}) {
+	var fields = []string{"type", "node", "probe", "old", "new"}
+	var sb strings.Builder
+	index := 0
+
+	sb.WriteString("SELECT * FROM EVENTS ")
+	if len(params) == 0 {
+		return sb.String(), args
+	}
+	sb.WriteString("WHERE ")
+
+	for _, key := range fields {
+		if val, ok := params[key]; ok {
+			sb.WriteString(fmt.Sprintf("%s = ?", key))
+			args = append(args, val)
+		}
+		if index < len(params)-1 {
+			sb.WriteString("AND ")
+		}
+		index++
+	}
+	return sb.String(), args
+}
+
+// sqlEvent defines an sql event row.
+type sqlEvent struct {
+	ID        int            `db:"id"`
+	Timestamp time.Time      `db:"timestamp"`
+	EventType string         `db:"type"`
+	Node      sql.NullString `db:"node"`
+	Probe     sql.NullString `db:"probe"`
+	Old       sql.NullString `db:"oldState"`
+	New       sql.NullString `db:"newState"`
+}
+
+// newSQLEvent constructs a new sqlEvent from the provided TimelineEvent.
+func newSQLEvent(event *pb.TimelineEvent) (row sqlEvent, err error) {
+	row = sqlEvent{Timestamp: event.GetTimestamp().ToTime()}
+	switch t := event.GetData().(type) {
+	case *pb.TimelineEvent_ClusterRecovered:
+		row.EventType = clusterRecoveredType
+	case *pb.TimelineEvent_ClusterDegraded:
+		row.EventType = clusterDegradedType
+	case *pb.TimelineEvent_NodeAdded:
+		e := event.GetNodeAdded()
+		row.EventType = nodeAddedType
+		row.Node = sql.NullString{String: e.GetNode(), Valid: true}
+	case *pb.TimelineEvent_NodeRemoved:
+		e := event.GetNodeRemoved()
+		row.EventType = nodeRemovedType
+		row.Node = sql.NullString{String: e.GetNode(), Valid: true}
+	case *pb.TimelineEvent_NodeRecovered:
+		e := event.GetNodeRecovered()
+		row.EventType = nodeRecoveredType
+		row.Node = sql.NullString{String: e.GetNode(), Valid: true}
+	case *pb.TimelineEvent_NodeDegraded:
+		e := event.GetNodeDegraded()
+		row.EventType = nodeDegradedType
+		row.Node = sql.NullString{String: e.GetNode(), Valid: true}
+	case *pb.TimelineEvent_ProbeSucceeded:
+		e := event.GetProbeSucceeded()
+		row.EventType = probeSucceededType
+		row.Node = sql.NullString{String: e.GetNode(), Valid: true}
+		row.Probe = sql.NullString{String: e.GetProbe(), Valid: true}
+	case *pb.TimelineEvent_ProbeFailed:
+		e := event.GetProbeFailed()
+		row.EventType = probeFailedType
+		row.Node = sql.NullString{String: e.GetNode(), Valid: true}
+		row.Probe = sql.NullString{String: e.GetProbe(), Valid: true}
+	default:
+		return row, trace.NotFound("unknown event type %v", t)
```

Comment:
I think `trace.BadParameter` is better suited here. Also, add `%T` to log the type as well.